File: user-management/keycloak-gateway/src/main/kotlin/co/nilin/opex/auth/gateway/extension/RegistrationOpexCaptcha.java
/*
 * Copyright 2016 Red Hat, Inc. and/or its affiliates
 * and other contributors as indicated by the @author tags.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package co.nilin.opex.auth.gateway.extension;

import org.apache.http.client.methods.CloseableHttpResponse;
import org.apache.http.client.methods.HttpGet;
import org.apache.http.client.utils.URIBuilder;
import org.apache.http.impl.client.CloseableHttpClient;
import org.jboss.logging.Logger;
import org.keycloak.Config;
import org.keycloak.authentication.FormAction;
import org.keycloak.authentication.FormActionFactory;
import org.keycloak.authentication.FormContext;
import org.keycloak.authentication.ValidationContext;
import org.keycloak.connections.httpclient.HttpClientProvider;
import org.keycloak.events.Details;
import org.keycloak.events.Errors;
import org.keycloak.forms.login.LoginFormsProvider;
import org.keycloak.models.*;
import org.keycloak.models.utils.FormMessage;
import org.keycloak.provider.ConfiguredProvider;
import org.keycloak.provider.ProviderConfigProperty;
import org.keycloak.services.ServicesLogger;
import org.keycloak.services.validation.Validation;

import javax.ws.rs.core.MultivaluedMap;
import java.util.ArrayList;
import java.util.List;

/**
 * @author <a href="mailto:bill@burkecentral.com">Bill Burke</a>
 * @version $Revision: 1 $
 */
public class RegistrationOpexCaptcha implements FormAction, FormActionFactory, ConfiguredProvider {
    public static final String CAPTCHA_ANSWER = "captcha-answer";
    public static final String PROVIDER_ID = "registration-opex-captcha-action";

    private static final Logger logger = Logger.getLogger(RegistrationOpexCaptcha.class);

    @Override
    public String getDisplayType() {
        return "Opex-Captcha";
    }

    @Override
    public String getReferenceCategory() {
        return null;
    }

    @Override
    public boolean isConfigurable() {
        return false;
    }

    private static final AuthenticationExecutionModel.Requirement[] REQUIREMENT_CHOICES = {
            AuthenticationExecutionModel.Requirement.REQUIRED,
            AuthenticationExecutionModel.Requirement.DISABLED
    };

    @Override
    public AuthenticationExecutionModel.Requirement[] getRequirementChoices() {
        return REQUIREMENT_CHOICES;
    }

    @Override
    public void buildPage(FormContext context, LoginFormsProvider form) {
    }

    @Override
    public void validate(ValidationContext context) {
        MultivaluedMap<String, String> formData = context.getHttpRequest().getDecodedFormParameters();
        List<FormMessage> errors = new ArrayList<>();
        boolean success = false;
        context.getEvent().detail(Details.REGISTER_METHOD, "form");

        String captcha = formData.getFirst(CAPTCHA_ANSWER);
        if (!Validation.isBlank(captcha)) {
            success = validateOpexCaptcha(context, captcha);
        }
        if (success) {
            context.success();
        } else {
            errors.add(new FormMessage(null, "opexCaptchaFailed"));
            formData.remove(CAPTCHA_ANSWER);
            context.error(Errors.INVALID_REGISTRATION);
            context.validationError(formData, errors);
            context.excludeOtherErrors();
        }
    }

    protected boolean validateOpexCaptcha(ValidationContext context, String captcha) {
        boolean success = false;
        try {
            CloseableHttpClient httpClient = (CloseableHttpClient) context.getSession().getProvider(HttpClientProvider.class).getHttpClient();
            HttpGet post = new HttpGet(new URIBuilder("https://captcha:8080").addParameter("proof", captcha).build());
            try (CloseableHttpResponse response = httpClient.execute(post)) {
                if (response.getStatusLine().getStatusCode() / 500 == 5)
                    throw new IllegalStateException("Failed to verify captcha");
                success = response.getStatusLine().getStatusCode() / 100 == 2;
            }
        } catch (Exception e) {
            ServicesLogger.LOGGER.error(e);
        }
        return success;
    }

    @Override
    public void success(FormContext context) {
    }

    @Override
    public boolean requiresUser() {
        return false;
    }

    @Override
    public boolean configuredFor(KeycloakSession session, RealmModel realm, UserModel user) {
        return true;
    }

    @Override
    public void setRequiredActions(KeycloakSession session, RealmModel realm, UserModel user) {
    }

    @Override
    public boolean isUserSetupAllowed() {
        return false;
    }

    @Override
    public void close() {
    }

    @Override
    public FormAction create(KeycloakSession session) {
        return this;
    }

    @Override
    public void init(Config.Scope config) {
    }

    @Override
    public void postInit(KeycloakSessionFactory factory) {
    }

    @Override
    public String getId() {
        return PROVIDER_ID;
    }

    @Override
    public String getHelpText() {
        return "Enables Opex-Captcha test.";
    }

    @Override
    public List<ProviderConfigProperty> getConfigProperties() {
        return new ArrayList<>();
    }
}

File: user-management/keycloak-gateway/src/main/kotlin/co/nilin/opex/auth/gateway/extension/RegistrationOpexCaptcha.kt
package co.nilin.opex.auth.gateway.extension

import org.apache.http.client.methods.HttpGet
import org.apache.http.client.utils.URIBuilder
import org.apache.http.impl.client.CloseableHttpClient
import org.keycloak.Config
import org.keycloak.authentication.FormAction
import org.keycloak.authentication.FormActionFactory
import org.keycloak.authentication.FormContext
import org.keycloak.authentication.ValidationContext
import org.keycloak.connections.httpclient.HttpClientProvider
import org.keycloak.events.Details
import org.keycloak.events.Errors
import org.keycloak.forms.login.LoginFormsProvider
import org.keycloak.models.*
import org.keycloak.models.utils.FormMessage
import org.keycloak.provider.ConfiguredProvider
import org.keycloak.provider.ProviderConfigProperty
import org.keycloak.services.ServicesLogger
import org.keycloak.services.validation.Validation

class RegistrationOpexCaptcha : FormAction, FormActionFactory, ConfiguredProvider {
    override fun getDisplayType(): String {
        return "Opex-Captcha"
    }

    override fun getReferenceCategory(): String? {
        return null
    }

    override fun isConfigurable(): Boolean {
        return false
    }

    override fun getRequirementChoices(): Array<AuthenticationExecutionModel.Requirement> {
        return REQUIREMENT_CHOICES
    }

    override fun buildPage(context: FormContext, form: LoginFormsProvider) {}

    override fun validate(context: ValidationContext) {
        val formData = context.httpRequest.decodedFormParameters
        val errors: MutableList<FormMessage> = ArrayList()
        var success = false
        context.event.detail(Details.REGISTER_METHOD, "form")
        val captcha = formData.getFirst(CAPTCHA_ANSWER)
        if (!Validation.isBlank(captcha)) {
            success = validateOpexCaptcha(context, captcha)
        }
        if (success) {
            context.success()
        } else {
            errors.add(FormMessage(null, "opexCaptchaFailed"))
            formData.remove(CAPTCHA_ANSWER)
            context.error(Errors.INVALID_REGISTRATION)
            context.validationError(formData, errors)
            context.excludeOtherErrors()
        }
    }

    override fun success(context: FormContext) {}

    override fun requiresUser(): Boolean {
        return false
    }

    override fun configuredFor(session: KeycloakSession, realm: RealmModel, user: UserModel): Boolean {
        return true
    }

    override fun setRequiredActions(session: KeycloakSession, realm: RealmModel, user: UserModel) {}

    override fun isUserSetupAllowed(): Boolean {
        return false
    }

    override fun close() {}

    override fun create(session: KeycloakSession): FormAction {
        return this
    }

    override fun init(config: Config.Scope) {}

    override fun postInit(factory: KeycloakSessionFactory) {}

    override fun getId(): String {
        return PROVIDER_ID
    }

    override fun getHelpText(): String {
        return "Enables Opex-Captcha test."
    }

    override fun getConfigProperties(): List<ProviderConfigProperty> {
        return ArrayList()
    }

    private fun validateOpexCaptcha(context: ValidationContext, captcha: String?): Boolean {
        var success = false
        try {
            val httpClient = context.session.getProvider(
                HttpClientProvider::class.java
            ).httpClient as CloseableHttpClient
            val xForwardedFor = context.httpRequest.httpHeaders.getRequestHeader("X-Forwarded-For")
            val proof = "$captcha-${xForwardedFor.first()}"
            val post = HttpGet(URIBuilder("http://captcha:8080").addParameter("proof", proof).build())
            httpClient.execute(post).use { response ->
                check(response.statusLine.statusCode / 500 != 5) { "Could not connect to Opex-Captcha service." }
                success = response.statusLine.statusCode / 100 == 2
            }
        } catch (e: Exception) {
            ServicesLogger.LOGGER.error(e)
        }
        return success
    }

    companion object {
        const val CAPTCHA_ANSWER = "captcha-answer"
        const val PROVIDER_ID = "registration-opex-captcha-action"
        private val REQUIREMENT_CHOICES = arrayOf(
            AuthenticationExecutionModel.Requirement.REQUIRED, AuthenticationExecutionModel.Requirement.DISABLED
        )
    }
}
